app/gui2/src/util/database/reactiveDb.ts

```diff
@@ -0,0 +1,197 @@
+  }
+}
+
+/**
+ * A function type representing an indexer for a `ReactiveDb`.
+ *
+ * An `Indexer` takes a key-value pair from the `ReactiveDb` and produces an array of index key-value pairs,
+ * defining how the input key and value maps to keys and values in the index.
+ *
+ * @param key - The key from the `ReactiveDb`.
+ * @param value - The value from the `ReactiveDb`.
+ *
+ * @returns An Array representing multiple key-value pair(s) ([IK, IV]) for the index.
+ */
+export type Indexer<K, V, IK, IV> = (key: K, value: V) => [IK, IV][]
+
+/**
+ * Provides automatic indexing for a `ReactiveDb` instance.
+ * Utilizes both forward and reverse mapping for efficient lookups and reverse lookups.
+ *
+ * @typeParam K - The key type of the ReactiveDb.
+ * @typeParam V - The value type of the ReactiveDb.
+ * @typeParam IK - The key type of the index.
+ * @typeParam IV - The value type of the index.
+ */
+export class ReactiveIndex<K, V, IK, IV> {
+  /** Forward map from index keys to a set of index values */
+  forward: Map<IK, Set<IV>>
+  /** Reverse map from index values to a set of index keys */
+  reverse: Map<IV, Set<IK>>
+  /** Collections of effects to sync data between ReactiveDB and ReactiveIndex */
+  effects: LazySyncEffectSet
+
+  /**
+   * Constructs a new ReactiveIndex for the given ReactiveDb and an indexer function.
+   *
+   * @param db - The ReactiveDb to index.
+   * @param indexer - The indexer function defining how db keys and values map to index keys and values.
+   */
+  constructor(db: ReactiveDb<K, V>, indexer: Indexer<K, V, IK, IV>) {
+    this.forward = reactive(new Map())
+    this.reverse = reactive(new Map())
+    this.effects = new LazySyncEffectSet()
+    db.on('entryAdded', (key, value, onDelete) => {
+      const stopEffect = this.effects.lazyEffect((onCleanup) => {
+        const keyValues = indexer(key, value)
+        keyValues.forEach(([key, value]) => this.writeToIndex(key, value))
+        onCleanup(() => keyValues.forEach(([key, value]) => this.removeFromIndex(key, value)))
+      })
+      onDelete(() => stopEffect())
+    })
+  }
+
+  /**
+   * Adds a new key-value pair to the forward and reverse index.
+   *
+   * @param key - The key to add to the index.
+   * @param value - The value to associate with the key.
+   */
+  writeToIndex(key: IK, value: IV): void {
+    const forward = setIfUndefined(this.forward, key, () => new Set())
+    forward.add(value)
+    const reverse = setIfUndefined(this.reverse, value, () => new Set())
+    reverse.add(key)
+  }
+
+  /**
+   * Removes a key-value pair from the forward and reverse index.
+   *
+   * @param key - The key to remove from the index.
+   * @param value - The value associated with the key.
+   */
+  removeFromIndex(key: IK, value: IV): void {
+    const remove = <K, V>(map: Map<K, Set<V>>, key: K, value: V) => {
+      map.get(key)?.delete(value)
+    }
+    remove(this.forward, key, value)
+    remove(this.reverse, value, key)
```

Comment:
When an identical `[key, value]` pair is returned from two independent `indexer` invocations (e.g. on two different entries), and one of them is later deleted, its cleanup will cause the pair to be completely removed from index. Even if one of the indexers that returned it is still "live". To prevent that, we can do one of two things:
- If that behavior is considered an indexer implementation bug, make sure it's discoverable (i.e. add an error log in `writeToIndex` if that pair is already in the index)
- Or, if it is a legitimately useful thing that can happen, add support for it by tracking the **count** of how many times any key,value pair has been indexed, then only clean it up when that count reaches 0.